comp/core/workloadmeta/collectors/internal/containerd/image_sbom_trivy.go

```diff
@@ -62,27 +82,63 @@ func (c *collector) startSBOMCollection(ctx context.Context) error {
 		}
 	}()
 
-	go c.startScanResultHandler(ctx, resultChan)
-
 	return nil
 }
 
+// startRetryLoop starts a loop that will retry SBOM generation for images that failed to generate a SBOM
+func (c *collector) startRetryLoop(ctx context.Context, resultChan chan<- sbom.ScanResult) {
+	for {
+		imgID, shutdown := c.queue.Get()
+		if shutdown {
+			log.Debugf("shutdown sbom retry queue")
+			return
+		}
+		id, ok := imgID.(string)
+		if !ok {
+			log.Errorf("expected string, got %T", imgID)
+			c.queue.Forget(imgID)
+			c.queue.Done(imgID)
+			continue
+		}
+		imgMeta, err := c.store.GetImage(id)
+		if err != nil {
+			log.Debugf("image %s not found in store", id)
+			c.queue.Forget(imgID)
+			c.queue.Done(imgID)
+			continue
+		}
+		log.Debugf("retrying SBOM generation for image %s (id: %s)", imgMeta.Name, imgMeta.ID)
+		c.sendScanRequest(ctx, imgMeta, resultChan)
+		// c.queue.Done() cannot be called here as it as the sbom is being processed
+	}
+}
+
 // handleEventBundle handles ContainerImageMetadata set events for which no SBOM generation attempt was done.
 func (c *collector) handleEventBundle(ctx context.Context, eventBundle workloadmeta.EventBundle, resultChan chan<- sbom.ScanResult) {
 	eventBundle.Acknowledge()
 	for _, event := range eventBundle.Events {
 		image := event.Entity.(*workloadmeta.ContainerImageMetadata)
 
-		if image.SBOM.Status != workloadmeta.Pending {
-			// A generation attempt has already been done. In that case, it should go through the retry logic.
-			// We can't handle them here otherwise it would keep retrying in a while loop.
-			log.Debugf("Image: %s/%s (id %s) SBOM already available", image.Namespace, image.Name, image.ID)
-			continue
+		switch image.SBOM.Status {
+		case workloadmeta.Success:
+			log.Debugf("SBOM available for image: %s/%s (id %s) available", image.Namespace, image.Name, image.ID)
+			// Forget and Done are safe to call even if the image is not in the retry queue.
+			c.queue.Forget(image.ID)
+			c.queue.Done(image.ID)
+		case workloadmeta.Failed:
+			// We can't store the image itself because it is updated at every iteration so the queue won't be able to find it.
+			c.queue.AddRateLimited(image.ID)
+			// Done needs to be called after processing the image. It is only processed once the scanner returns a result.
+			c.queue.Done(image.ID)
```

Comment:
I am not sure how it would work. From the doc:
https://github.com/kubernetes/client-go/blob/f2d91322bf55f2be90e300c4845d365470cd125a/util/workqueue/queue.go#L220-L222

And the function that marks an item dirty is `AddRateLimited`. It also means that we `retry` before knowing the result.